### ISSUE TITLE
Better handling of extreme long lines when justify?=true

### ISSUE THREAD

**Issue by lucywang000** [kkinnear/zprint#179]
I'd like to ask something about `{justify? true}`, since being able to justify binding pairs is one of the main reasons why in #178 I asked if it's possible to "my-wrapper-macro" to have its first argument recognized as bindings. 

Is it possible to only justify some of the bindings that are close enough, and leave out those would make the line too long because two many spaces would be inserted? For instance, the original code could be:
```clojure
(let [a 1
      bb 2
      ccc 3
      dddd 4
      {:keys [foo bar baz bark key1 key2 key3 key4] :as spam} 1
      bb 2
      ccc 3
      dddd 4])
```

Right now zprint would format it to:
```clojure
(let [a                                                       1
      bb                                                      2
      ccc                                                     3
      dddd                                                    4
      {:keys [foo bar baz bark key1 key2 key3 key4] :as spam} 1
      bb                                                      2
      ccc                                                     3
      dddd                                                    4])
```
Code like this where the binding pairs are two far away causes back and forth movements of the eye when our sight move from line to line, and could be tiring, and it's the major reason that I turned off "justify?" earlier yesterday after playing with it.

 So some middle ground could be better:
```clojure
(let [a    1
      bb   2
      ccc  3
      dddd 4
      {:keys [foo bar baz bark key1 key2 key3 key4] :as spam} 1
      bb   2
      ccc  3
      dddd 4])
```
IMO it's apparently better than the previous two format: nothing could be done the long destructuring line, but the lines before and after it could be formatted for better reading. So maybe we could add some option like justify-limit=10 which would not justify a line if taking care of it would cause more than 10 extra spaces would be inserted to other lines.

Another idea is to break the destructuring into multiple lines and have the last line justified:

```clojure
(let [a                            1
      bb                           2
      ccc                          3
      dddd                         4
      {:keys [foo bar baz
              bark key1 key2
              key3 key4] :as spam} 1
      bb                           2
      ccc                          3
      dddd                         4])
```

**Comment by kkinnear** [comment]
Thanks for thinking about this. I don't personally use `{:style justified}`, though there are places where it would be nice, probably in part because of the issue you point out with highly disparate line lengths.

These are both interesting ideas.  The "middle ground" would be easier to implement with the current approach to justification, as opposed to the "Another idea".  At present justification is used if the justified output is not longer  (i.e., doesn't take more lines) than the unjustified output (or maybe not "much" longer, I don't remember the exact details).  zprint places a lot of importance on getting the most code readably presented in given amount of vertical space -- since no matter how big your screen is, having more of your code visible is almost always better than not.  So, while I kind of like the "Another idea" approach, the "middle ground" one would probably not require nearly the rework as the "Another idea" approach and might be nearly as good. 

I'll take a look at the code and see what I can come up with.  Thanks again for the idea!

**Comment by kkinnear** [comment]
Fixed in `1.1.2`.

This was a great idea!  I reworked all of the justification code to use the variance of the left-hand sides of pairs when making justification decisions.  It will now leave out up to two rows that cause the variance to be over the configured maximum, in an attempt to bring the variance below that maximum.  This is now the default approach.  For example:
```
zprint.core=> (print i179)
(let [a 1
      bb 2
      ccc 3
      dddd 4
      {:keys [foo bar baz bark key1 key2 key3 key4] :as spam} 1
      bb 2
      ccc 3
      dddd 4])
nil
zprint.core=> (czprint i179 {:parse-string? true :style :justified})
(let [a    1
      bb   2
      ccc  3
      dddd 4
      {:keys [foo bar baz bark key1 key2 key3 key4], :as spam} 1
      bb   2
      ccc  3
      dddd 4])
```
Because of this conversation I have several other ideas for how to improve justification, but they are much larger implementation efforts that will have to wait for a bit.  At least this one makes things look a lot better.  

Thanks again for the idea!
